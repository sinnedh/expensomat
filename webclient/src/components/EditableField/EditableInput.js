import React from 'react'

class EditableInput extends React.Component {
  state = {
    editValue: this.props.value,
    editMode: false
  }

  handleClickEdit(event) {
    event.preventDefault()
    this.setState({editValue: this.props.value, editMode: true})
  }

  handleClickCancel(event) {
    event.preventDefault()
    this.setState({editValue: '', editMode: false})
  }

  handleClickSave(event) {
    event.preventDefault()
    this.setState({editMode: false})
    this.props.onClickSave(this.state.editValue)
  }

  handleChangeValue(event) {
    this.setState({editValue: event.target.value})
  }

  render() {
    if(this.state.editMode) {
      return (
        <span>
          <input className="value" type="text" value={this.state.editValue} onChange={e => this.handleChangeValue(e)} />
          <input className="save-button" type="button" value="Save" onClick={e => this.handleClickSave(e)} />
          <input className="cancel-button" type="button" value="Cancel" onClick={e => this.handleClickCancel(e)} />
        </span>
      )
    } else {
      return (
        <span>
          {this.props.value}
          <input className="edit-button" type="button" value="Edit" onClick={e => this.handleClickEdit(e)} />
        </span>
      )
    }
  }
}

export default EditableInput
