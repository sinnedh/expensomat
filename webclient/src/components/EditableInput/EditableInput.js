import React from 'react'

class EditableInput extends React.Component {
  state = {
    initialValue: this.props.value,
    value: this.props.value,
    editMode: false
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.value === '') {
      this.setState({initialValue: nextProps.value, value: nextProps.value})
    }
  }

  handleClickEdit(event) {
    event.preventDefault()
    this.setState({editMode: true})
  }

  handleClickCancel(event) {
    event.preventDefault()
    this.setState({value: this.state.initialValue, editMode: false})
  }

  handleClickSave(event) {
    event.preventDefault()
    this.setState({editMode: false})
    this.props.onClickSave(this.state.value)
  }

  handleChangeValue(event) {
    this.setState({value: event.target.value})
  }

  render() {
    if(this.state.editMode) {
      return (
        <span>
          <input className="value" type="text" value={this.state.value} onChange={e => this.handleChangeValue(e)} />
          <input className="save-button" type="button" value="Save" onClick={e => this.handleClickSave(e)} />
          <input className="cancel-button" type="button" value="Cancel" onClick={e => this.handleClickCancel(e)} />
        </span>
      )
    } else {
      return (
        <span>
          {this.state.value}
          <input className="edit-button" type="button" value="Edit" onClick={e => this.handleClickEdit(e)} />
        </span>
      )
    }
  }
}

export default EditableInput
