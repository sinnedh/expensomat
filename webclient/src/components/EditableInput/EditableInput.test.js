import React from 'react'
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import EditableInput from './EditableInput'

it('renders when in show mode', () => {
  const wrapper = shallow(<EditableInput value={"Hello"} />)
  wrapper.setState({editMode: false})
  expect(wrapper).toMatchSnapshot()
})

it('renders when in edit mode', () => {
  const wrapper = shallow(<EditableInput value={"Hello"} />)
  wrapper.setState({editMode: true})
  expect(wrapper).toMatchSnapshot()
})

it('calls onUpdateInput when input is updated', () => {
  const wrapper = shallow(<EditableInput value={"Hello"} />)
  wrapper.setState({editMode: true})
  wrapper.find('input.value').simulate('change', { target: { value: 'Hell' } })
  expect(wrapper.find('input.value').props().value).toBe('Hell')
})

it('shows edit mode when edit button is clicked', () => {
  const wrapper = shallow(<EditableInput value={"Hello"} />)
  wrapper.find('input.edit-button').simulate('click', { preventDefault() {} })
  expect(wrapper.find('input.value')).toHaveLength(1)
})

it('leaves edit mode when cancel button is clicked', () => {
  const wrapper = shallow(<EditableInput value={"Hello"} />)
  wrapper.setState({editMode: true})
  wrapper.find('input.cancel-button').simulate('click', { preventDefault() {} })
  expect(wrapper.find('input.value')).toHaveLength(0)
})

it('save button calls onClickSave with input value', () => {
  const onClickSave = jest.fn();
  const wrapper = shallow(
    <EditableInput onClickSave={onClickSave} value={"Hello"} />
  )

  wrapper.setState({editMode: true})
  wrapper.find('input.save-button').simulate('click', { preventDefault() {} })
  expect(onClickSave).toBeCalledWith('Hello')
  expect(wrapper.find('input.value')).toHaveLength(0)
})
