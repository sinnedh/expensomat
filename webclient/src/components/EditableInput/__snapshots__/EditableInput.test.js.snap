// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when in edit mode 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableInput
    value="Hello"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="value"
          onChange={[Function]}
          type="text"
          value="Hello"
/>,
        <input
          className="save-button"
          onClick={[Function]}
          type="button"
          value="Save"
/>,
        <input
          className="cancel-button"
          onClick={[Function]}
          type="button"
          value="Cancel"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "value",
          "onChange": [Function],
          "type": "text",
          "value": "Hello",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "save-button",
          "onClick": [Function],
          "type": "button",
          "value": "Save",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "cancel-button",
          "onClick": [Function],
          "type": "button",
          "value": "Cancel",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="value"
            onChange={[Function]}
            type="text"
            value="Hello"
/>,
          <input
            className="save-button"
            onClick={[Function]}
            type="button"
            value="Save"
/>,
          <input
            className="cancel-button"
            onClick={[Function]}
            type="button"
            value="Cancel"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "value",
            "onChange": [Function],
            "type": "text",
            "value": "Hello",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "save-button",
            "onClick": [Function],
            "type": "button",
            "value": "Save",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "cancel-button",
            "onClick": [Function],
            "type": "button",
            "value": "Cancel",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders when in show mode 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableInput
    value="Hello"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "Hello",
        <input
          className="edit-button"
          onClick={[Function]}
          type="button"
          value="Edit"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      "Hello",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "edit-button",
          "onClick": [Function],
          "type": "button",
          "value": "Edit",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Hello",
          <input
            className="edit-button"
            onClick={[Function]}
            type="button"
            value="Edit"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        "Hello",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "edit-button",
            "onClick": [Function],
            "type": "button",
            "value": "Edit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
